crates/wasmtime/src/store.rs

```diff
@@ -2180,12 +2195,19 @@ impl Drop for StoreOpaque {
             let ondemand = OnDemandInstanceAllocator::default();
             for instance in self.instances.iter_mut() {
                 if instance.ondemand {
-                    ondemand.deallocate(&mut instance.handle);
+                    ondemand.deallocate_module(&mut instance.handle);
                 } else {
-                    allocator.deallocate(&mut instance.handle);
+                    allocator.deallocate_module(&mut instance.handle);
+                }
+            }
+            ondemand.deallocate_module(&mut self.default_caller);
+
+            #[cfg(feature = "component-model")]
+            {
+                for _ in self.component_instances.iter() {
```

Comment:
here we're using the component-instance array just for its length -- could we store a counter rather than actual instance handles?